### DISCUSSION TITLE
Policy regarding disputed countries/territories

### DISCUSSION THREAD

**Discussion by MunyuShizumi** [dr5hn/countries-states-cities-database#327] [General]
(_Any specific references are provided only as examples for the sake of understanding/illustration, not in any political context!_)

A couple of questions regarding disputed countries/territories and such, mostly because I need some level of predictability for user-facing data (and it's extremely difficult to have a geopolitically neutral systematization).

---

### 1. When is a disputed country listed (only) as an independent country?
I see a few cases where not-(yet)-unanimously-recognized countries are listed as implicitly independent in this database. Ex. Kosovo (more recognized than not internationally), while not officially recognized by Serbia, is listed only as a country.

Interestingly enough, ISO 3166 has yet to include Kosovo in their systematization (colloquially, the user-assigned XK code is often used). It's currently listed as an ISO 3166-2 subdivision of Serbia, with Kosovo's counties listed as second-level subdivisions.

---

### 2. When is a disputed country listed as both an independent country and a subdivision?
Taiwan is listed both as a separate country and a province of China. ISO 3166 follows the same logic, where it's listed as both "TW" and "CN-TW" under ISO 3166-1 and 3166-2 respectively, despite having comparatively limited international recognition. ISO 3166 generally seems to do this quite often, even for non-disputed territories (ex. unincorporated US territories and such). 

---

### 3. How is entity removal handled?
I'm not sure how Kosovo was or would have been categorized before (ISO 3166-2 lists it as a single first-level subdivision of Serbia and counties as second-level subdivisions), but some restructuring has or would have happened when it was recategorized as a country. I'll use current ISO 3166-2 codes to illustrate various possible changes (I know, ISO 3166-2 has parent-child subdivision relations, this database is a strict country->state->city system, meh), but do interpret these as integer IDs:
- RS-KM removed entirely as a state, XK added as a new country. Subdivisions RS-25/26/27/28/29 marked as XK first-level children. I.e. country ID is new, state IDs have **not** changed, parent country IDs reassigned to newly created country.
- RS-KM removed entirely as a state, XK added as a new country. Subdivisions RS-25/26/27/28/29 removed entirely. New subdivisions added for XK (possibly with same/similar names). I.e. country ID is new, old state IDs **entirely removed**, new state IDs added (with new country ID as parent).

Overall, removal of any entities creates a problem when ex. a hypothetical user selects a state that later becomes an independent country as their location. Since states and countries are strictly separate entities rather than flexible subdivisions, the following problems might occur:
- Database is left with old IDs which are no longer valid country/state foreign keys, either throwing an exception or making it impossible to dereference the original country/state names.
- If states keep their old IDs, strict parent-child constraints are violated where states are left referencing the old country.

---

Sorry about the wall of text, I'd also like to avoid this problem, but it presents a major future issue for my project(s). I'd happily contribute to help ameliorate this somehow (and have a few suggestions on how to possibly resolve it), but I feel it's currently in limbo policy-wise so wanted to check what your thoughts were first.

(And just to explicitly state it: great work on this project, seems rather promising, it's just that the world is sometimes more complex than we data engineers prefer. ._.)

**Comment by dr5hn** [comment]
Hi @MunyuShizumi,

Thanks for putting this up. Really appreciate. 🎉 

Let me think about this and come up with some concrete solution.

I know, There's a major concern with Foreign key relationships when children are deleted or a parent is moved to another parent.
